### PR TITLE
feat: Add resource name resolutions for ELB, DCS, VPC

### PR DESCRIPTION
**Pull request by DrDJIng**
This PR is for resource name resolution for ELB, DCS, and VPC resources.

ELB didn't seem to be functioning, so I replaced the the client with the new openstack v3 client, which successfully retrieves the resource names.

DCS was a simple addition of a DCS openstack client.

VPC, however, required some workarounds. There was no working VPC client, so I use the vpcep client and replaced the endpoints and resourcebases so that it points at the correct endpoint for bandwidths and public ips. This also required another change to pass the entire config into the FetchResourceIdToNameMapping function so that the projectId can be used in the endpoint remapping.

If the openstack clients then gain vpc client or bandwidth/publicip clients, this can be simply replaced.

### CODE REVIEW

**Comment by zeljkobekcic** [comment]
Sorry for the delay, I will take a look at it now.

File: internal/otc_wrapper.go [hunk 1]
```diff
@@ -212,6 +214,110 @@ func (c *OtcWrapper) GetDdsIdNameMapping() (map[string]string, error) {
 	return result, nil
 }
 
+func (c *OtcWrapper) GetDcsIdNameMapping() (map[string]string, error) {
+	opts := golangsdk.EndpointOpts{Region: "eu-de"}
```
Comment (zeljkobekcic) [inline internal/otc_wrapper.go:218]: it would be better here to access `c.region` instead of hardcoded "eu-de".

File: internal/otc_wrapper.go [hunk 2]
```diff
@@ -212,6 +214,110 @@ func (c *OtcWrapper) GetDdsIdNameMapping() (map[string]string, error) {
...
+		if instance.Name != "" {
+			result[instance.InstanceID] = instance.Name
+		}
+	}
+
+	return result, nil
+}
+
+func (c *OtcWrapper) GetVpcIdNameMapping(projectId string) (map[string]string, error) {
+	opts := golangsdk.EndpointOpts{Region: "eu-de"}
```
Comment (zeljkobekcic) [inline internal/otc_wrapper.go:242]: see above

File: internal/otc_wrapper.go [hunk 3]
```diff
@@ -212,6 +214,110 @@ func (c *OtcWrapper) GetDdsIdNameMapping() (map[string]string, error) {
...
+	vpcClient, err := openstack.NewVpcEpV1(c.providerClient, opts)
+	vpcClient.Endpoint = strings.Replace(vpcClient.Endpoint, "vpcep", "vpc", 1)
+	vpcClient.Endpoint = strings.Replace(vpcClient.Endpoint, fmt.Sprintf("%s/", projectId), "", 1)
+	vpcClient.ResourceBase = vpcClient.Endpoint
+
```
Comment (zeljkobekcic) [inline internal/otc_wrapper.go:247]: Digging around, I thing the otc go lib does not provide this api access?

Comment (zeljkobekcic) [reply internal/otc_wrapper.go:247]: I feel like we should create a issue on the otc go lib for this.

File: main.go [hunk 4]
```diff
@@ -82,10 +82,10 @@ func collectMetricsInBackground(config internal.ConfigStruct, logger internal.IL
 	}()
 }
 
-func FetchResourceIdToNameMapping(client *internal.OtcWrapper, namespaces []string) (map[string]string, error) {
+func FetchResourceIdToNameMapping(client *internal.OtcWrapper, config internal.ConfigStruct) (map[string]string, error) {
```
Comment (zeljkobekcic) [inline main.go:85]: I feel like we should maybe adapt the otc wrapper because this behaviour could be required for other namespaces too.

**Review by zeljkobekcic — COMMENTED** [review]
Some parts require a little bit discussion. Especially the VPC client thing. Generally looks good. But I am not the biggest fan of the manual endpoint manipulation.

File: main.go (hunk 4 above)
Comment (DrDJIng) [reply main.go:85]: As in fold the configstruct into the wrapper?

**Comment by DrDJIng** [comment]
Sorry it's been a while from me as well!

Yeah, I admit the endpoint manipulation isn't the most elegant solution but I needed to get it to work somehow. I think solving the problem in the otc gopher base would be the best solution, but it's a workaround in the meantime.

Perhaps an issue and a comment with the issue link in the code itself could be used as a reminder to fix it once the original problem is solved?

It depends on whether expanded resource name resolution functionality is a priority feature or not, I guess?

File: internal/otc_wrapper.go (hunk 3 above)
Comment (zeljkobekcic) [reply internal/otc_wrapper.go:247]: I will create one shortly for that.

**Comment by zeljkobekcic** [comment]
I am fine with this PR, just replace the hard coded "eu-de". Once this is done I am gonna merge this.

This PR here asks about the client for VPC: https://github.com/opentelekomcloud/gophertelekomcloud/issues/626

Sorry for the delayed answer.

**Comment by DrDJIng** [comment]
I hope I didn't miss anything! Definitely wouldn't have thought that NewNetworkV1 was the way to go.

**Review by zeljkobekcic — APPROVED** [review]

**Comment by zeljkobekcic** [comment]
No worries, this is good! I will merge it now. Just changed the title a bit.

**Comment by zeljkobekcic** [comment]
The feature looks great and I would love to merge it. You just need to merge main into your branch @DrDJIng

**Review by zeljkobekcic — APPROVED** [review]

**Pull request merged**